### PR TITLE
generate kube with volumes

### PR DESCRIPTION
**Pull request by haircommander**
Add volume functionality for generate kube, including adding the fields for containers, as well as the pod. Test added as well
Fixes: https://github.com/containers/libpod/issues/2303

### CODE REVIEW

File: libpod/kube.go [hunk 1]
```diff
@@ -190,56 +203,58 @@ func addContainersToPodObject(containers []v1.Container, podName string) *v1.Pod
...
 	if len(c.config.Spec.Linux.Devices) > 0 {
 		// TODO Enable when we can support devices and their names
 		devices, err := generateKubeVolumeDeviceFromLinuxDevice(c.Spec().Linux.Devices)
 		if err != nil {
-			return kubeContainer, err
+			return kubeContainer, kubeVolumes, err
 		}
 		kubeContainer.VolumeDevices = devices
-		return kubeContainer, errors.Wrapf(define.ErrNotImplemented, "linux devices")
+		return kubeContainer, kubeVolumes, errors.Wrapf(define.ErrNotImplemented, "linux devices")
```
Comment (mheon) [inline libpod/kube.go]: Wait, what's going on here? We're getting devices... and then throwing an error that they're not supported?

Comment (rhatdan) [reply libpod/kube.go]: Should this just be a warning?

Comment (haircommander) [reply libpod/kube.go]: `TODO Enable when we can support devices and their names` I assume the translation isn't clear for libpod devices to kube devices?

File: libpod/kube.go [hunk 2]
```diff
@@ -325,36 +340,81 @@ func libpodMaxAndMinToResourceList(c *Container) (v1.ResourceList, v1.ResourceLi
 	return maxResources, minResources
 }
 
-func generateKubeVolumeMount(hostSourcePath string, mounts []specs.Mount) (v1.VolumeMount, error) {
+func generateKubeVolumeMount(hostSourcePath string, mounts []specs.Mount) (v1.VolumeMount, v1.Volume, error) {
```
Comment (mheon) [inline libpod/kube.go]: Might want to just pass the container in here - would be easier to work with. You loop through UserVolumes, identify named volume vs normal volume, and then add based on that. Should remove the need to deduplicate, for example.

Example code lives in https://github.com/containers/libpod/blob/master/libpod/container_inspect.go#L360-L426

Might be worth making a helper somewhere in libpod that does that job - retrieves all user-added volumes into volumes and named volumes.

Comment (haircommander) [reply libpod/kube.go]: The dedup is for containers that share volumes, not necessarily for volumes one container has. the scope in `generateKubeVolumeMount` is too narrow (container level, not pod level) to have this fix the dup problem

Comment (mheon) [reply libpod/kube.go]: Ahh, makes sense.

I think we should still avoid the current method of parsing all mounts in the spec... That's asking for trouble in some ways

Comment (haircommander) [reply libpod/kube.go]: an added bonus is we aren't needlessly doing an n^2 slice lookup anymore!

File: test/e2e/generate_kube_test.go [hunk 3]
```diff
@@ -207,4 +207,19 @@ var _ = Describe("Podman generate kube", func() {
...
+	It("podman generate kube with volume", func() {
+		vol1 := filepath.Join(podmanTest.TempDir, "vol-test1")
+		err := os.MkdirAll(vol1, 0755)
+		Expect(err).To(BeNil())
+
+		session1 := podmanTest.Podman([]string{"run", "-d", "--pod", "new:test1", "-v", vol1 + ":/volume/:z", "alpine", "top"})
+		session1.WaitWithDefaultTimeout()
+		Expect(session1.ExitCode()).To(Equal(0))
+
+		kube := podmanTest.Podman([]string{"generate", "kube", "test1"})
```
Comment (mheon) [inline test/e2e/generate_kube_test.go]: Can we play the container we generate here with `play kube` and verify it has a volume present?

Comment (haircommander) [reply test/e2e/generate_kube_test.go]: fixed!

**Comment by mheon** [comment]
Changes LGTM

**Comment by haircommander** [comment]
@baude @jwhonce PTAL

**Comment by mheon** [comment]
@baude @jwhonce @rhatdan @TomSweeneyRedHat @vrothberg  Can we get this landed? I'm messing around in Inspect and I don't want to make merge conflicts...

**Comment by TomSweeneyRedHat** [comment]
Changes LGTM, if you're happy with the one question you had @mheon, feel free to merge it.
At some point, I'd like to see some expanded doc covering this new functionality.

**Pull request merged**